### ISSUE TITLE
pyscal version - ModuleNotFoundError: No module named 'pyscal'

### ISSUE THREAD

**Issue by Mateus-M-Reis** [pyiron/pyiron_atomistics#1024]
pip installation failing.

I'm on a Xubuntu **22.04.2 LTS** x86_64 without GPU.
Trying to install `pyiron 0.4.7`.

I am using [pyenv ](https://github.com/pyenv/pyenv) and [virtualenvwrapper](https://github.com/python-virtualenvwrapper/virtualenvwrapper) to manage my python virtual environments:
```
mkvirtualenv pyiron            # python 3.10
pip install pyiron             # also tryied with pip install pyiron --pre
```

Moving on to download the pyiron resources:
```
python
>>> import pyiron
```

I receive the following error:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/machinedr/.virtualenvs/pyiron/lib/python3.11/site-packages/pyiron/__init__.py", line 4, in <module>
    from pyiron_atomistics.project import Project
  File "/home/machinedr/.virtualenvs/pyiron/lib/python3.11/site-packages/pyiron_atomistics/__init__.py", line 4, in <module>
    from pyiron_atomistics.project import Project
  File "/home/machinedr/.virtualenvs/pyiron/lib/python3.11/site-packages/pyiron_atomistics/project.py", line 30, in <module>
    from pyiron_atomistics.lammps.potential import LammpsPotentialFile
  File "/home/machinedr/.virtualenvs/pyiron/lib/python3.11/site-packages/pyiron_atomistics/lammps/__init__.py", line 1, in <module>
    from pyiron_atomistics.lammps.potential import view_potentials, list_potentials
  File "/home/machinedr/.virtualenvs/pyiron/lib/python3.11/site-packages/pyiron_atomistics/lammps/potential.py", line 13, in <module>
    from pyiron_atomistics.atomistics.structure.atoms import Atoms
  File "/home/machinedr/.virtualenvs/pyiron/lib/python3.11/site-packages/pyiron_atomistics/atomistics/structure/atoms.py", line 15, in <module>
    from structuretoolkit import (
  File "/home/machinedr/.virtualenvs/pyiron/lib/python3.11/site-packages/structuretoolkit/__init__.py", line 5, in <module>
    from structuretoolkit.analyse.pyscal import (
  File "/home/machinedr/.virtualenvs/pyiron/lib/python3.11/site-packages/structuretoolkit/analyse/pyscal.py", line 7, in <module>
    import pyscal.core as pc
ModuleNotFoundError: No module named 'pyscal'
```

So I try to install `pyscal`:
```
>>> exit()
pip install pyscal
```
the following heappens:

```
Collecting pyscal
  Using cached pyscal-0.10.3-py2.py3-none-any.whl (138 kB)
Requirement already satisfied: matplotlib in ./.virtualenvs/pyiron/lib/python3.11/site-packages (from pyscal) (3.7.1)
Requirement already satisfied: numpy in ./.virtualenvs/pyiron/lib/python3.11/site-packages (from pyscal) (1.24.3)
Collecting openpyxl
  Using cached openpyxl-3.1.2-py2.py3-none-any.whl (249 kB)
Requirement already satisfied: pandas in ./.virtualenvs/pyiron/lib/python3.11/site-packages (from pyscal) (2.0.1)
Collecting scipy<1.9.3
  Using cached scipy-1.9.2-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (33.4 MB)
Collecting xlrd
  Using cached xlrd-2.0.1-py2.py3-none-any.whl (96 kB)
Requirement already satisfied: contourpy>=1.0.1 in ./.virtualenvs/pyiron/lib/python3.11/site-packages (from matplotlib->pyscal) (1.0.7)
Requirement already satisfied: cycler>=0.10 in ./.virtualenvs/pyiron/lib/python3.11/site-packages (from matplotlib->pyscal) (0.11.0)
Requirement already satisfied: fonttools>=4.22.0 in ./.virtualenvs/pyiron/lib/python3.11/site-packages (from matplotlib->pyscal) (4.39.3)
Requirement already satisfied: kiwisolver>=1.0.1 in ./.virtualenvs/pyiron/lib/python3.11/site-packages (from matplotlib->pyscal) (1.4.4)
Requirement already satisfied: packaging>=20.0 in ./.virtualenvs/pyiron/lib/python3.11/site-packages (from matplotlib->pyscal) (23.1)
Requirement already satisfied: pillow>=6.2.0 in ./.virtualenvs/pyiron/lib/python3.11/site-packages (from matplotlib->pyscal) (9.5.0)
Requirement already satisfied: pyparsing>=2.3.1 in ./.virtualenvs/pyiron/lib/python3.11/site-packages (from matplotlib->pyscal) (3.1.0b1)
Requirement already satisfied: python-dateutil>=2.7 in ./.virtualenvs/pyiron/lib/python3.11/site-packages (from matplotlib->pyscal) (2.8.2)
Collecting et-xmlfile
  Using cached et_xmlfile-1.1.0-py3-none-any.whl (4.7 kB)
Requirement already satisfied: pytz>=2020.1 in ./.virtualenvs/pyiron/lib/python3.11/site-packages (from pandas->pyscal) (2023.3)
Requirement already satisfied: tzdata>=2022.1 in ./.virtualenvs/pyiron/lib/python3.11/site-packages (from pandas->pyscal) (2023.3)
Requirement already satisfied: six>=1.5 in ./.virtualenvs/pyiron/lib/python3.11/site-packages (from python-dateutil>=2.7->matplotlib->pyscal) (1.16.0)
Installing collected packages: xlrd, scipy, et-xmlfile, openpyxl, pyscal
  Attempting uninstall: scipy
    Found existing installation: scipy 1.10.1
    Uninstalling scipy-1.10.1:
      Successfully uninstalled scipy-1.10.1
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
pyiron-atomistics 0.2.66 requires scipy>=1.10.1, but you have scipy 1.9.2 which is incompatible.
structuretoolkit 0.0.1 requires scipy>=1.10.0, but you have scipy 1.9.2 which is incompatible.
Successfully installed et-xmlfile-1.1.0 openpyxl-3.1.2 pyscal-0.10.3 scipy-1.9.2 xlrd-2.0.1
```
Which doesn't work:
```
python
>>> import pyiron
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/machinedr/.virtualenvs/pyiron/lib/python3.10/site-packages/pyiron/__init__.py", line 4, in <module>
    from pyiron_atomistics.project import Project
  File "/home/machinedr/.virtualenvs/pyiron/lib/python3.10/site-packages/pyiron_atomistics/__init__.py", line 4, in <module>
    from pyiron_atomistics.project import Project
  File "/home/machinedr/.virtualenvs/pyiron/lib/python3.10/site-packages/pyiron_atomistics/project.py", line 30, in <module>
    from pyiron_atomistics.lammps.potential import LammpsPotentialFile
  File "/home/machinedr/.virtualenvs/pyiron/lib/python3.10/site-packages/pyiron_atomistics/lammps/__init__.py", line 1, in <module>
    from pyiron_atomistics.lammps.potential import view_potentials, list_potentials
  File "/home/machinedr/.virtualenvs/pyiron/lib/python3.10/site-packages/pyiron_atomistics/lammps/potential.py", line 13, in <module>
    from pyiron_atomistics.atomistics.structure.atoms import Atoms
  File "/home/machinedr/.virtualenvs/pyiron/lib/python3.10/site-packages/pyiron_atomistics/atomistics/structure/atoms.py", line 15, in <module>
    from structuretoolkit import (
  File "/home/machinedr/.virtualenvs/pyiron/lib/python3.10/site-packages/structuretoolkit/__init__.py", line 5, in <module>
    from structuretoolkit.analyse.pyscal import (
  File "/home/machinedr/.virtualenvs/pyiron/lib/python3.10/site-packages/structuretoolkit/analyse/pyscal.py", line 7, in <module>
    import pyscal.core as pc
ModuleNotFoundError: No module named 'pyscal.core'

```

More info:
```
pip list | grep pyiron
pyiron             0.4.7
pyiron-atomistics  0.2.66
pyiron-base        0.5.37

```

**Comment by jan-janssen** [comment]
@Mateus-M-Reis The easiest way to install pyiron is using the `conda-forge` community repositories: 
```
conda install -c conda-forge pyiron
```

Unfortunately, the `pyscal` package is not available on pypi https://github.com/pyscal/pyscal still you can install it using `pip`:
```
pip install git+https://github.com/srmnitc/pyscal.git
```

**Comment by jan-janssen** [comment]
@srmnitc Any chance you can publish pyscal on pypi ?

**Comment by Mateus-M-Reis** [comment]
`pyscal 0.10.3` is on pipy.
Running 
```
pip install pyscal
pip install pyiron
```
produces the following:
```
...
Installing collected packages: scandir, pyfiglet, py-cpuinfo, msgpack, mpmath, urllib3, typing-extensions, traitlets, tqdm, threadpoolctl, tenacity, tabulate, sympy, spglib, smmap, scipy, ruamel.yaml.clib, PyYAML, Pygments, pycparser, psutil, pint, pathlib2, palettable, numexpr, networkx, monty, MarkupSafe, latexcodec, joblib, idna, h5py, greenlet, future, dill, defusedxml, cython, colorama, charset-normalizer, certifi, blosc2, bcrypt, uncertainties, tables, SQLAlchemy, seekpath, scikit-learn, ruamel.yaml, requests, pydantic, pybtex, plotly, jinja2, h5io, gitdb, cffi, pynacl, pyfileindex, phonopy, mendeleev, gitpython, cryptography, ase, paramiko, pysqa, pyiron-base, emmet-core, mp-api, pymatgen, aimsgb, structuretoolkit, pyiron-atomistics, pyiron
  Attempting uninstall: scipy
    Found existing installation: scipy 1.9.2
    Uninstalling scipy-1.9.2:
      Successfully uninstalled scipy-1.9.2
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
pyscal 0.10.3 requires scipy<1.9.3, but you have scipy 1.10.1 which is incompatible.
...
```
I wasn't able to find which version of `scipy` does pyiron requires, there's no `requirements.txt` file. By the output above I conclude that it is `1.10.1`.

**Comment by niklassiemer** [comment]
Yes, it is set in https://github.com/pyiron/pyiron_atomistics/blob/69f83f9210a8255179697af17f7e847d92d5d5c3/setup.py#L57
We always use the most recent versions if possible, however, this is not needed to be a real no-go for a lower version, since we not necessarily also use the newest features.

**Comment by srmnitc** [comment]
> `pyscal 0.10.3` is on pipy. Running
> 
> ```
> pip install pyscal
> pip install pyiron
> ```
> 


The pyscal which is available on pypi is a different tool, albeit with the same name. As Jan mentions, conda is the preferred method of installation. If you want to install with pip, you have to use the following command:
```
pip uninstall pyscal
pip install git+https://github.com/pyscal/pyscal.git
```
The dependency issues are also coming from the wrong pyscal package.
@jan-janssen I will creating a pip package.

**Comment by Mateus-M-Reis** [comment]
Ok. Sorry about that.
Running the aforementioned commands do install pyscal, but I cannot get pyiron to install bacause a series of packages not being installed:
```
pip install git+https://github.com/pyscal/pyscal.git
...
pip install pyiron
...
python
...
>>> import pyiron
Traceback (most recent call last):
...
ModuleNotFoundError: No module named 'pybind11'
>>> exit()
pip install git+https://github.com/pybind/pybind11.git
python
...
>>> import pyiron
Traceback (most recent call last):
...
ModuleNotFoundError: No module named 'sqsgenerator'
exit()
pip install git+https://github.com/dgehringer/sqsgenerator.git
...
Building wheels for collected packages: sqsgenerator
  Building wheel for sqsgenerator (setup.py) ... error
  error: subprocess-exited-with-error
...
          raise CalledProcessError(retcode, cmd)
      subprocess.CalledProcessError: Command '['cmake', '/tmp/pip-req-build-u_xs_02o', '-DCMAKE_BUILD_TYPE=Release', '-DCMAKE_LIBRARY_OUTPUT_DIRECTORY_RELEASE=/tmp/pip-req-build-u_xs_02o/build/lib.linux-x86_64-cpython-310/sqsgenerator/core', '-DCMAKE_ARCHIVE_OUTPUT_DIRECTORY_RELEASE=build/temp.linux-x86_64-cpython-310', '-DPython3_EXECUTABLE=/home/machinedr/.virtualenvs/pyiron/bin/python', '-DUSE_MPI=OFF', '-DCMAKE_CXX_FLAGS_RELEASE=']' returned non-zero exit status 1.
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for sqsgenerator
  Running setup.py clean for sqsgenerator
Failed to build sqsgenerator
ERROR: Could not build wheels for sqsgenerator, which is required to install pyproject.toml-based projects
```
Anaconda consumes too much RAM to create an environment (compared with venv) so I avoid it is this machine with 8GB RAM. But right now there is no way around I guess. I'll go with miniconda, and when `pyscal` become available on pipy I'll try the process again to see what heappens.
Thanks for your help guys.

**Comment by jan-janssen** [comment]
There is https://github.com/conda-forge/miniforge - which uses `mamba` rather than `conda` and usually works much faster.

**Comment by jan-janssen** [comment]
I guess the general note we have to work on is reducing the dependency footprint of pyiron.

**Comment by Mateus-M-Reis** [comment]
> There is https://github.com/conda-forge/miniforge - which uses `mamba` rather than `conda` and usually works much faster.

Thanks for the sugestion. I'll give it a try.
 
Edit:
I tried it, but unfortunately without success. It looks like impossible to install `pyiron` with only 8GB RAM using anaconda, even with `miniforge/mamba`.

**Comment by jan-janssen** [comment]
@Mateus-M-Reis Is this any specific system? Apple m2 or similar?

**Comment by Mateus-M-Reis** [comment]
> @Mateus-M-Reis Is this any specific system? Apple m2 or similar?

No.
It's a Xubuntu desktop with Kernel 5.15.0-71-generic with a 
AMD Ryzen 5 3400G (8) @ 3.700GHz overclocked (4000 GHz) and 
AMD ATI Radeon Vega Series / Radeon Vega Mobile integrated graphics.

miniforge with mamba is faster, but the problem is the amount of RAM that anaconda consumes in the latest stage of the installation of packages. Generally this problem heappens when I try to install a lot of packages in the same environment (opencv, tensorflow, jupyter...) but just by installing pyiron my RAM burts and the process is killed.

Installing pyiron in a fresh environment:

![image](https://user-images.githubusercontent.com/42076854/234992488-124a8c71-9846-41d6-867f-d3f4ced6a581.png)

**Comment by srmnitc** [comment]
> @srmnitc Any chance you can publish pyscal on pypi ?

pyscal is now available on pypi as `pyscal2`, so one could do `pip install pyscal2`. @jan-janssen since it has C++ parts, a number of wheels have to built, I would think its better to keep the conda package as it is, and not built from pypi.

**Comment by jan-janssen** [comment]
Great work @srmnitc I guess we should also create a new Conda package named `pyscal2` and then link the existing package to the new one.

**Comment by jan-janssen** [comment]
I created the corresponding pull request https://github.com/conda-forge/staged-recipes/pull/22709 - @srmnitc Can you comment on it so that the conda-forge team sees you agree with being listed as maintainer.

**Comment by liamhuber** [comment]
> I tried it, but unfortunately without success. It looks like impossible to install `pyiron` with only 8GB RAM using anaconda, even with `miniforge/mamba`.

Hi Mateus,

I haven't done a fresh install for a long time, but I'm able to update pyiron ok on my old 2013 MacBook with 8GB ram. I have shared some of your experience though, as if I try to update the pyiron env from inside pycharm (perhaps even just with pycharm open), it typically hangs on the solve indefinitely. Shutting down my IDE and everything else that's consuming ram and doing the updates directly in the terminal with the conda CLI has always given me perfectly reasonable performance. Not sure this will
solve things for a fresh install but I hope it helps!

**Comment by liamhuber** [comment]
> > @Mateus-M-Reis Is this any specific system? Apple m2 or similar?
> 
> 
> 
> No.
> 
> It's a Xubuntu desktop with Kernel 5.15.0-71-generic with a 
> 
> AMD Ryzen 5 3400G (8) @ 3.700GHz overclocked (4000 GHz) and 
> 
> AMD ATI Radeon Vega Series / Radeon Vega Mobile integrated graphics.
> 
> 
> 
> miniforge with mamba is faster, but the problem is the amount of RAM that anaconda consumes in the latest stage of the installation of packages. Generally this problem heappens when I try to install a lot of packages in the same environment (opencv, tensorflow, jupyter...) but just by installing pyiron my RAM burts and the process is killed.
> 
> 
> 
> Installing pyiron in a fresh environment:
> 
> 
> 
> ![image](https://user-images.githubusercontent.com/42076854/234992488-124a8c71-9846-41d6-867f-d3f4ced6a581.png)
> 
> 

This is a serious problem if it's general and we should figure out how to reduce our dependency stack if that's the case. But in the meantime another workaround might be to conda install a number of the dependencies manually, and install pyiron component-wise, ie first base, then atomistics, etc. But this is absolutely just a stopgap and not a real solution!

**Comment by Mateus-M-Reis** [comment]
> > I tried it, but unfortunately without success. It looks like impossible to install `pyiron` with only 8GB RAM using anaconda, even with `miniforge/mamba`.
> 
> Hi Mateus,
> 
> I haven't done a fresh install for a long time, but I'm able to update pyiron ok on my old 2013 MacBook with 8GB ram. I have shared some of your experience though, as if I try to update the pyiron env from inside pycharm (perhaps even just with pycharm open), it typically hangs on the solve indefinitely. Shutting down my IDE and everything else that's consuming ram and doing the updates directly in the terminal with the conda CLI has always given me perfectly reasonable performance. Not sure this will solve things for a fresh install but I hope it helps!

I didn't work. Maybe because part of my RAM is being used esclusively to caching, as you can see from the prints above. I don't know how it works I just choose de default option when installing my OS.
I tried with just the terminal open, after killing some processes which were consuming RAM. 
I remember being able to install pyiron long time ago (1-2 years), on Elementary OS, which is heavier than Xubuntu.

> > > @Mateus-M-Reis Is this any specific system? Apple m2 or similar?
> > 
> > 
> > No.
> > It's a Xubuntu desktop with Kernel 5.15.0-71-generic with a
> > AMD Ryzen 5 3400G (8) @ 3.700GHz overclocked (4000 GHz) and
> > AMD ATI Radeon Vega Series / Radeon Vega Mobile integrated graphics.
> > miniforge with mamba is faster, but the problem is the amount of RAM that anaconda consumes in the latest stage of the installation of packages. Generally this problem heappens when I try to install a lot of packages in the same environment (opencv, tensorflow, jupyter...) but just by installing pyiron my RAM burts and the process is killed.
> > Installing pyiron in a fresh environment:
> > ![image](https://user-images.githubusercontent.com/42076854/234992488-124a8c71-9846-41d6-867f-d3f4ced6a581.png)
> 
> This is a serious problem if it's general and we should figure out how to reduce our dependency stack if that's the case. But in the meantime another workaround might be to conda install a number of the dependencies manually, and install pyiron component-wise, ie first base, then atomistics, etc. But this is absolutely just a stopgap and not a real solution!
 
I don't think installing the dependencies manually will make a difference.
As far as I am aware, conda dependency solving is somewhat cummulative. I remember having these problems whenever [sympy](https://github.com/sympy/sympy) (which is a **pyiron** dependency) was present in an environment. I installed **jupyterlab** and **sympy** in a fresh environment and then conda would struggle to solve the dependencies for the nest package whichever it was.

**Comment by liamhuber** [comment]
> I didn't work... I tried with just the terminal open, after killing some processes which were consuming RAM.

> I don't think installing the dependencies manually will make a difference. ... As far as I am aware, conda dependency solving is somewhat cummulative...

I can't say for sure one way or the other that there is an extra overhead on install, but my gut certainly agrees that the solve is the expensive part.

In this case I'm really surprised that your machine doesn't manage it with 8GB. My machine floats around 3GB usage for OS and basic programs, so with 5GB I get pretty fast solves. PyCharm eats another ~2.5GB, and when I leave that open and drop 5GB down to 2.5GB, then I absolutely see it destroy solves. 

> I remember having these problems whenever [sympy](https://github.com/sympy/sympy) (which is a pyiron dependency) was present in an environment.

If we can track it down to a few problematic packages and find alternatives, that would be super. I don't know exactly what the RAM threshold is for installing pyiron successfully, but obviously we want it as low as possible.

(As far as I can see, sympy is only used by pyiron_continuum -- for the fenics jobs, so actually it's my fault the dependency is there, but my work with the FEM stuff was pretty basic -- so I don't think this is the trouble maker in your case.)

**Comment by Mateus-M-Reis** [comment]
That problem I mentioned looks had been solved now. I installed `sympy` and some packages
and everything worked fine.

One a new fresh environment, I installed all packages in `pyiron-atomistic/setup.py` individually through conda-forge, they also ran smoothly (including `pyiron_base==0.5.37`), except by the following: `mendeleev==0.13.0` and `pandas==2.0.0`, which failed in the initial solve, outputing the following lines:
```
Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Collecting package metadata (repodata.json): done
Solving environment: done
```
I proceeded to install pyiron. I received the following lines:
```
Collecting package metadata (current_repodata.json): done
Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: failed with repodata from current_repodata.json, will retry with next repodata source.
Collecting package metadata (repodata.json): done
Solving environment: done
```
And everything ran fine.
`conda install -c conda-forge pyiron` automatically installs `nglview`, `sphinxdft-data` and `sqsgenerator`. I also managed to install `lammps` and `gpaw` in the same environment.

Thanks very much for your help guys. I'll program some molecules now!
